lib/spack/spack/solver/concretize.lp

```diff
@@ -1010,23 +997,24 @@ variant_is_propagated(PackageNode, Variant) :-
 % 1. The same flag type is not set on this node
 % 2. This node has the same compiler as the propagation source
 
-propagated_flag(node(PackageID, Package), node_flag(FlagType, Flag), SourceNode) :-
-  propagate(node(PackageID, Package), node_flag(FlagType, Flag), _),
-  not attr("node_flag_set", node(PackageID, Package), FlagType, _),
+propagated_flag(node(PackageID, Package), node_flag(FlagType, Flag, FlagGroup, Source), SourceNode) :-
+  propagate(node(PackageID, Package), node_flag(FlagType, Flag, FlagGroup, Source), _),
+  % Disable to pass propagate_and_spec_dep, propagate_and_pkg_dep
+  % and propagate_and_require
+  % not attr("node_flag_set", node(PackageID, Package), FlagType, _, _, _),
   % Same compiler as propagation source
   node_compiler(node(PackageID, Package), CompilerID),
   node_compiler(SourceNode, CompilerID),
-  attr("propagate", SourceNode, node_flag(FlagType, Flag), _),
+  attr("propagate", SourceNode, node_flag(FlagType, Flag, FlagGroup, Source), _),
```

Comment:
(question) I found it interesting that this was expressed as an `attr` vs. as:

`propagate(SourceNode, node_flag(FlagType, Flag, FlagGroup, Source), _)`

I see that:

* `asp.py` generates `propagate` as an attribute like `attr(propagate...)`
* `concretize.lp` generates a `propagate()` predicate from this
* All other rules in `concretize.lp` interact with this information via `propagate()` vs. `attr(propagate...)`

Why not use `propagate()` here?

Also, why doesn't `propagate()` encode edge info? As written, this seems susceptible to undesired associations:

```
  v
 / \ 
w   y  # ^w cflags==g, ^y cflags==g
|   |
x   z

Say x, y, z have the same compiler and w has a different compiler:
as encoded, flag propagation info can "jump" across branches.
```

Reply:
Only the sources of the propagation have the:
```
attr("propagate", Node, PropagatedAttribute)
```
Internally the propagation uses `propagate` facts.

Reply:
* I assume your response is to the first part of my comment (not addressing the concern about the lack of edge info in `propagate()`)
* Even so, it does not conflict with what I say: this line could use `propagate()` rather than `attr("propagate"...` (and IMO it would be worthwhile because it initially puzzled me why it differed from all the other references)

Reply:
The `propagate` rules just move some facts down the DAG. They're simple rule and propagate "dumbly" so that complexity is kept O(edges). The propagated facts should contain enough information for a node to decide locally whether a fact in `propagate` should have any effect on its `attr` or not.

`attr("propagate", ...)` is currently used to kick-off propagation, so it's only on the roots. We can use any other mechanism to kick-off `propagate` rules, but why?

Reply:
> We can use any other mechanism to kick-off propagate rules, but why?

I'm not against the mechanism for kicking it off

The general logic for dependencies like `X -> Y -> Z` is

* `attr("propagate")` on X generates `propagate()` on X
* `propagate()` on X applies transitively to Y/Z

Everything in this lineage (including the root) has `propagate()`, so I wondered why this constraint is conditioning on `attr("propagate")`.

Does it not seem strange that we have all of `propagate(X/Y/Z)`, but that we effectively ask "if Z has `propagate()`, and if we initialized propagation on X..."?

One reason I can think of *not* to swap to `propagate()` is that this may lead to more comparisons (because, as you pointed out, `attr("propagate"...` is just on the roots). Although that's not clear from an initial reading.

Overall:

* I understand how this works
* I thought it was confusing
* I wondered if you thought it was too
* It's not critical IMO to replace `propagate()` with `attr("propagate")`
* However, whether you do or don't, there is a potential issue for "branch jumping" (described in detail in 1st comment)